core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java

```diff
@@ -1133,11 +1133,14 @@ public boolean supportsMissingColumnsOnInsert(Session session, TableHandle table
     }
 
     @Override
-    public Optional<ConnectorOutputMetadata> finishInsert(Session session, InsertTableHandle tableHandle, Collection<Slice> fragments, Collection<ComputedStatistics> computedStatistics)
+    public Optional<ConnectorOutputMetadata> finishInsert(Session session, InsertTableHandle tableHandle, List<TableHandle> sourceTableHandles, Collection<Slice> fragments, Collection<ComputedStatistics> computedStatistics)
     {
         CatalogHandle catalogHandle = tableHandle.getCatalogHandle();
         ConnectorMetadata metadata = getMetadata(session, catalogHandle);
-        return metadata.finishInsert(session.toConnectorSession(catalogHandle), tableHandle.getConnectorHandle(), fragments, computedStatistics);
+        List<ConnectorTableHandle> sourceConnectorHandles = sourceTableHandles.stream()
+                .map(TableHandle::getConnectorHandle)
```

Comment:
The list of handles should be filtered to make sure they match this connector instance
```java
.filter(handle -> handle.getCatalogHandle().equals(catalogHandle))
```
The engine never mixes handles between connector instances.